lib/EnzymeCore/src/EnzymeCore.jl

```diff
@@ -506,28 +506,128 @@ function autodiff_thunk end
+of its constituent parts is guaranteed to be inactive (non-differentiable): reuse `prev`s
+instance (the default) or make a copy.
+
+The argument `runtime_inactive` specifies whether each constituent type is checked for being
+guaranteed inactive at runtime for every call to `make_zero`, or if this can be checked once
+at compile-time and reused across multiple calls to `make_zero` and related functions (the
+default). Runtime checks are necessary to pick up recently added methods to
+`EnzymeRules.inactive_type`, but may incur a significant performance penalty and is usually
+not needed unless `EnzymeRules.inactive_type` is extended interactively for types that have
+previously been passed to `make_zero` or related functions.
+
+Extending this method for custom types is rarely needed. If you implement a new type, such
+as a GPU array type, for which `make_zero` should directly invoke `zero` rather than
+iterate/broadcast when the eltype is scalar, it is sufficient to implement `Base.zero` and
+make sure your type subtypes `DenseArray`. (If subtyping `DenseArray` is not appropriate,
+extend [`EnzymeCore.isvectortype`](@ref) directly instead.)
 """
 function make_zero end
 
 """
-    make_zero!(val::T, seen::IdSet{Any}=IdSet())::Nothing
+    make_zero!(val::T, [seen::IdDict], ::Val{runtime_inactive}=Val(false))::Nothing
+
+Recursively set a variable's differentiable values to zero. Only applicable for types `T`
+that are mutable or hold all differentiable values in mutable storage (e.g.,
+`Tuple{Vector{Float64}}` qualifies but `Tuple{Float64}` does not). The recursion skips over
+parts of `val` that are guaranteed to be inactive.
 
-Recursively set a variables differentiable fields to zero. Only applicable for mutable types `T`.
+The argument `runtime_inactive` specifies whether each constituent type is checked for being
+guaranteed inactive at runtime for every call to `make_zero!`, or if this can be checked once
+at compile-time and reused across multiple calls to `make_zero!` and related functions (the
+default). Runtime checks are necessary to pick up recently added methods to
+`EnzymeRules.inactive_type`, but may incur a significant performance penalty and is usually
+not needed unless `EnzymeRules.inactive_type` is extended interactively for types that have
+previously been passed to `make_zero!` or related functions.
+
+Extending this method for custom types is rarely needed. If you implement a new mutable
+type, such as a GPU array type, for which `make_zero!` should directly invoke
+`fill!(x, false)` rather than iterate/broadcast when the eltype is scalar, it is sufficient
+to implement `Base.zero`, `Base.fill!`, and make sure your type subtypes `DenseArray`. (If
+subtyping `DenseArray` is not appropriate, extend [`EnzymeCore.isvectortype`](@ref) directly
+instead.)
 """
 function make_zero! end
 
 """
-    make_zero(prev::T)
+    isvectortype(::Type{T})::Bool
```

Comment:
This is a new API? It will need a version bump for EnzymeCore.

Reply:
We should discuss. The reason I put these helpers in EnzymeCore rather than keeping them internal was that the StaticArrays extension needed to add a method, so I figured there's a chance others might have to do the same for their custom types. However, subtyping `DenseArray` (and `AbstractFloat` if that ever becomes relevant) should almost always be sufficient. Either way, the point is only to make these extensible in package extensions. I don't think anyone should ever have to _call_ them.

Reply:
However, the vector space wrapper functionality I've built on top of this (which will be a separate PR) will probably involve a new type in EnzymeCore, so if that gets accepted there will have to be a new EnzymeCore release anyway

Reply:
Yeah, absolutely. I think this is the right place to add them, and EnzymeCore is basically meant for people to be able to extend things without having to bite the load time bullet that is Enzyme.

Reply:
Btw. names are infinitely bikesheddable, both in this case and elsewhere in the PR. My mindset working on this PR is to enable consistent treatment of arbitrary objects as vectors in a space spanned by the scalar (float) values reachable from the object, hence all the vector/scalar terminology, but I don't know if this works well or can be confusing as part of the public API.